Manager v2

dickermoshe opened the pull request:
This PR includes the following changes to the manager filter API.

### Remove `Value` Filter for Custom Type
Until now the manage would create 2 filters for a column that has a TypeConverter.
One that accepts the Custom Type and another that uses the raw value.
To avoid polluting the namespace of filterset, the ability to filter on a raw value will be removed.
It is extremely uncommon that a filter on this raw data would be performed, so I think it's worth it.
E.G 
```dart
f.category.status(Status.active)
f.category.statusValue(7) // Removed
```

### References don't use a callback
This will help with readability:
```dart
items.filter((f)=>f.category.name("Foo"))
```
instead of 
```dart
items.filter((f)=>f.category((f)=>f.name("Foo")))
```

### Remove `Id` filter for references
Until now, references columns would create 2 columns, one for the actual column, and one for the reference object.
We will remove the "Id" filter.
Before:
```dart
items.filter((f)=>f.categoryId(5))
```
After:
```dart
items.filter((f)=>f.category.id(5))
```
This looks like we are performing a join for no reason, but we aren't!
Under the hood no join is performed, a filter is applied to the actual category column!


## Other Changes
1. Docs for extensions have been removed until the internal manager API is more consistent. We don't want people writing code at this lower level.
2. Mark lower level classes as internal, but still expose them for the generated code. This will change once the lower level APIs are more stable.
3. Disable manager generation on older dart 2 code

dickermoshe commented:
@simolus3 Adding tests and came across a bug.
The following SQL is completely valid:
```sql
SELECT
  DISTINCT COUNT(*) AS "c0"
FROM
  "book_club"
  LEFT OUTER JOIN "person" "book_club__id__person__club" ON "book_club"."id" = "book_club__id__person__club"."club"
GROUP BY
  "book_club"."_rowid_"
HAVING
  COUNT(*) = ?;
```

However, I'm getting the completly wrong output from row.read:

```dart
/// Count the number of rows that would be returned by the built statement
  Future<int> count() async {
    final count = _tableAsTableInfo.rowId.count();
    final result =
        _buildSelectStatement(targetColumns: [count]) as _JoinedResult;
    print(result.statement.constructQuery().sql); // Added for debugging
    print(result.statement.constructQuery().boundVariables);
    return await result.statement
        .map((row) => row.read(count)!)
        .get()
        .then((value) => value.firstOrNull ?? 0);
  }
```
The above is returning whatever `?` is.
So if:
```
HAVING
  COUNT(*) = 54;
```
Then the above function will return 54. There aren't even 54 clubs!
If seems that `row.read(count)!` is reading the wrong count?!

Just want to ensure that I'm not doing anything wrong before opening a issue on drift

dickermoshe commented:
@simolus3 I tried debugging this myself, but I'm at a loss.

There are Book Clubs and Members.
I'm filtering book clubs with 3 members.
This SQL query returns 1 club, which is CORRECT!
```sql
SELECT
  "book_club"."id" AS "book_club.id",
  "book_club"."name" AS "book_club.name"
FROM
  "book_club"
  LEFT OUTER JOIN "person" "book_club__id__person__club" ON "book_club"."id" = "book_club__id__person__club"."club"
GROUP BY
  "book_club"."_rowid_"
HAVING
  COUNT(*) = ?;
```

HOWEVER, COUNTING DOESN'T!
```sql
SELECT
  DISTINCT COUNT(*) AS "c0"
FROM
  "book_club"
  LEFT OUTER JOIN "person" "book_club__id__person__club" ON "book_club"."id" = "book_club__id__person__club"."club"
GROUP BY
  "book_club"."_rowid_"
HAVING
  COUNT(*) = ?;
```

This will always return whatever the number filter was, in this case, it's returning 3! There are only 2 clubs in the database!
Even accessing the raw value, gets `{c0: 3}`

This is either a really bad bug in drift or I'm doing something wrong.

This is the test that's failing, It's mind boggling.

https://github.com/dickermoshe/drift/blob/21b7607a77bfeb50b8cb56119043a636d10fa74d/drift/test/manager/manager_referenced_filter_test.dart#L161-L173

dickermoshe commented:
This is how the django ORM does counts.
Can this be created using drift?
```dart
SELECT
  COUNT(*)
FROM
  (
    SELECT
      "api_product"."sku" AS "col1"
    FROM
      "api_product"
      LEFT OUTER JOIN "api_listing" ON ("api_product"."sku" = "api_listing"."product_id")
    GROUP BY
      1
    HAVING
      COUNT("api_listing"."id") = 3
  )
```

dickermoshe commented:
@simolus3 
As usual your awesome documentation made this easy!

dickermoshe commented:
@simolus3 
Counting can only really be done if there is a rowid or only 1 primary key.
Even ORMs that have been developed for the past 20 years don't support a database with more than 1 primary key.
We can either throw an error if they try to `count` (it's like that now) or ignore such tables when generating manager (makes more sense, less shoot-foot-gun issues).

What should be done?

simolus3 commented:
>  Counting can only really be done if there is a rowid or only 1 primary key.

To be sure I'm understanding this correctly, the problem is that we care about the amount of unique rows of one table in a select statement that may have additional joins?
That's not really obvious from the documentation of that method, which just claims to return the amount of rows that would be returned by the whole statement. If that's the number we actually care about, what's wrong with `SELECT count(*) FROM (the actual statement that would otherwise be generated)`?

If we need "amount of unique rows from table x appearing in a joined result set", couldn't we add all the primary key columns to the inner result set and then do a sum + count combination? E.g. if the primary key is the columns `foo`, `bar`, `baz`:

```
SELECT sum(*) FROM (SELECT count(*) FROM (inner query) GROUP BY foo, bar, baz)
```

Then the inner query gives us how many rows there are for each unique combination of primary key columns and the final `sum` gives us the total amount. But I also don't see how that information is relevant in most queries.

dickermoshe commented:
That's real smart.

On a side note, I was looking at other ORM designs and have decided that aggregates should be done in another method and filtering should focus on filtering.

```dart
// Class where aggregates are predefined, these will be added to the filterset
class CategoryAggregateModel extends ...{
  ...
  final itemCount = ...;
  ...
}

category.annotate(CategoryAggregateModel()).filter((f)=>f.itemCount(5))
```

This will reduce code duplication if we ever decide to add a way to return the aggregated values.

I'm going to revert back to this design for filtering reverse references
```dart
// Category with todo with id of 5
category.filter(f.todos((f)=>f.id(5))) // Remove .exist() or .any()
```

Aggregates will be for a later release, they add complexity to the point where my limited knowledge of SQL is starting to show.

dickermoshe commented:
I've followed your guidance on counting and have begun adding more and more tests, It's starting to come together nicely.

simolus3 commented:
Nice! Let me know if there's anything ready for me to take a more in-depth look at :)

dickermoshe committed:
refactor: mark manager api classes as @internal
refactor: rename `HasJoinBuilders` to Composable
refactor: remove Composer fields to seperate ComposerState class
fix: composers inherits parent composers - fixes nested joins
refactor: add _BaseColumnFilters to reduce code duplication on columns with type converters
docs: add more docstrings

dickermoshe committed:
chore: disable manager generation for dart 2 code

dickermoshe committed:
test: add much more tests to manager

dickermoshe committed:
chore: update manager writer for new internal manager api

dickermoshe committed:
chore: regenerate projects
chore: remove manager extensions from docs until later notice

dickermoshe committed:
fix: add missing import
fix: hide internal imports for legacy projects

dickermoshe commented:
The refactoring here is quite signifying.
LMK if there is anything I can do to make your reviewing on this PR easier.

This PR is complete

simolus3 commented on docs/lib/snippets/dart_api/manager.dart:
```diff
@@ -210,61 +209,62 @@ extension ManagerExamples on AppDatabase {
-  db.managers.todoItems.orderBy((f) => f.contentThenCreatedAt());
-}
-// #enddocregion manager_custom_ordering
+// This documentation should added once the internal manager APIs are more stable
```
IMO we can leave them uncommented in this snippet file, as long as there's no snippet block on the markdown page referencing them they're not visible to users.

simolus3 commented on drift/lib/drift.dart:
```diff
@@ -20,8 +20,8 @@ export 'src/runtime/query_builder/query_builder.dart'
 export 'src/runtime/types/mapping.dart' hide BaseSqlType, UserDefinedSqlType;
-export 'src/runtime/manager/manager.dart'
-    hide JoinBuilder, HasJoinBuilders, Composer, BaseTableManager;
+// ignore: invalid_export_of_internal_element
```
If this is necessary, we should either make the affected elements non-`@internal` or indeed hide them.

There are a bunch of classes in drift that are only meant to be used by generated code. I don't have `@internal` annotations on them but instead mention this in their documentation comment.
Adding `@internal` on constructors or methods not used outside of `package:drift` would also work to hide them specifically.

simolus3 commented on drift/lib/src/runtime/manager/composable.dart:
````diff
@@ -50,29 +55,26 @@ class JoinBuilder {
 /// Example:
 /// ```dart
-/// categories.filter((f) => f.todos((f) => f.dueDate.isBefore(DateTime.now())))
+/// todos.filter((f) => f.category )
````
Can this be completed into a working filter (e.g. `(f) => f.category.id(3)`) so that the example makes a little more sense on its own?

simolus3 commented on drift/lib/src/runtime/manager/filter.dart:
```diff
@@ -118,201 +194,150 @@ extension StringFilters<T extends String> on ColumnFilters<String> {
-    return ComposableFilter(column.equals(_customTypeToSql(value)),
-        inverted: inverted);
-  }
+enum _Opperator { and, or }
```
nit, `_Operator` (or perhaps `_BooleanOperator`)

simolus3 commented on drift_dev/lib/src/backends/build/drift_builder.dart:
```diff
@@ -440,7 +440,8 @@ class _DriftBuildRun {
 
-    output.writeln('// ignore_for_file: type=lint');
+    output.writeln(
+        '// ignore_for_file: type=lint, invalid_use_of_internal_member');
```
It's a clever solution to the "members only supposed to be used by generated code" problem, but I think I still prefer the approach of mentioning this in their dartdoc without relying on this lint.

dickermoshe replied:
> but I think I still prefer the approach of mentioning this in their dartdoc without relying on this lint.

What do you mean by this?

I'm confused on how to proceed.

I would like it if changes to this low level API weren't classified as breaking changes.
When `@internal` is removed, test fail due to the `melos analyze` test

I've taken inspiration from Riverpod which uses this lint ignore on their generated code, however if you don't like it that perfectly fine.

Moving the `@internal` to the constructors does solve the ugly `export` -> `hide` issues, but the lint issues are still present.

How should I proceed?

simolus3 commented on drift_dev/lib/src/cli/commands/schema/generate_utils.dart:
```diff
@@ -167,5 +167,5 @@ class GenerateUtilsCommand extends Command {
   static final _dartfmt = DartFormatter();
   static const _prefix = '// GENERATED CODE, DO NOT EDIT BY HAND.\n'
-      '// ignore_for_file: type=lint';
+      '// ignore_for_file: type=lint, invalid_use_of_internal_member';
```
Same here

simolus3 commented on drift_dev/lib/src/writer/manager_writer.dart:
```diff
@@ -88,32 +94,59 @@ class _ReferencedFilterWriter extends _FilterWriter {
 
+  /// Whether this is a reverse reference
+  /// This is used to determine of the filter should
+  /// use a callback or return the related filterset directly
```
Let's be more specific here and mention that reverse references get a callback - an explanation for that would also be nice, I guess it's for improved readability in the forward direction whereas we want to be more explicit for backrefs?

simolus3 commented on extras/integration_tests/drift_testcases/lib/tests.dart:
```diff
@@ -1,6 +1,25 @@
+        ColumnFilters,
+        ColumnWithTypeConverterFilters,
+        TableManagerState,
+        RootTableManager;
```
And without the internal marker, re-exporting drift is easier than this :)

simolus3 commented on drift/lib/src/runtime/manager/composable.dart:
```diff
@@ -42,6 +42,11 @@ class JoinBuilder {
     return currentColumn.hashCode ^ referencedColumn.hashCode;
   }
 
+  @override
```
I've missed this in the previous PR. But while we're at it, can we change `hashCode` to use `Object.hash` instead of xors? I think `==` should also not be covariant and instead do a `other is JoinBuilder` check - otherwise it throws instead of returning false which is pretty weird.

simolus3 reviewed (commented):
Thanks! The API changes and refactorings look good to me (but I've yet to try them out). I just have some minor comments on the implementation.

dickermoshe committed:
style: more @internal to constructors
style: uncomment snippets
refactor: update JoinBuilder hashing logic
docs: explain revese reference filtering
fix: update generator not to ignore internal imports
chore: regenerate

dickermoshe committed:
fix: change `@internal` to doccomment

dickermoshe commented:
Let me know if there is anything else I can do...

dickermoshe committed:
export base table manager

simolus3 reviewed (approved):
Looks good to me, thanks again for all your hard work on this!

Is this ready to be released or do you have something in mind that should still be added or changed before a public version?

dickermoshe commented:
Yeah, this is ready for release.
